Initial commit for adding point clouds and labels

hchittanuru3 opened the pull request.

Commit:
Initial commit for adding point clouds and labels

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -84,17 +101,28 @@ def save_json_dict(json_fpath: Union[str, "os.PathLike[str]"], dictionary: Dict[
+	return log_ids
 
-def main(save_images: bool, save_poses: bool, save_calibration: bool):
+def main(save_images: bool, save_poses: bool, save_calibration: bool, save_cloud: bool, save_labels:bool):
```
i recommend we keep line length down to 80 chars so that it's a bit more readable:
https://google.github.io/styleguide/pyguide.html#32-line-length

Do you mind adding type hints as well? e.g. per https://google.github.io/styleguide/pyguide.html#221-type-annotated-code

Static analysis of types is possible with http://mypy-lang.org/
https://github.com/python/mypy

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -84,17 +101,28 @@ def save_json_dict(json_fpath: Union[str, "os.PathLike[str]"], dictionary: Dict[
 	""" """
 	# TFRECORD_DIR = 'VAL_TFRECORDS'
-	TFRECORD_DIR = 'TEST_TFRECORDS'
+	TFRECORD_DIR = '/srv/share3/hchittanuru3/waymo_training'
```
can these be argparse command line parameters? fine if your values are used as the argparse defaults

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -84,17 +101,28 @@ def save_json_dict(json_fpath: Union[str, "os.PathLike[str]"], dictionary: Dict[
 	# TFRECORD_DIR = 'VAL_TFRECORDS'
-	TFRECORD_DIR = 'TEST_TFRECORDS'
+	TFRECORD_DIR = '/srv/share3/hchittanuru3/waymo_training'
+	ARGO_WRITE_DIR = '/srv/share3/hchittanuru3/testing_waymo2argo'
```
same story here

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -84,17 +101,28 @@ def save_json_dict(json_fpath: Union[str, "os.PathLike[str]"], dictionary: Dict[
 	with open(json_fpath, "w") as f:
 		json.dump(dictionary, f)
 
+def get_log_id_from_files(files, record_dir):
```
type hints would be nice here. Do you mind also adding a 1-line docstring explaining the method's purpose, and then a bit of a longer docstring explaining the input string and output string format

let's use Google-style docstrings. Do you mind reading through this section?
https://google.github.io/styleguide/pyguide.html#381-docstrings

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -84,17 +101,28 @@ def save_json_dict(json_fpath: Union[str, "os.PathLike[str]"], dictionary: Dict[
+def get_log_id_from_files(files, record_dir):
+	log_ids = []
+	for file in files:
+		file = file.strip(record_dir)
```
what should `record_dir` look like?

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -107,22 +135,36 @@ def main(save_images: bool, save_poses: bool, save_calibration: bool):
+			# Reading lidar data and saving it point cloud format
+			(range_images,
+			camera_projections,
+			range_image_top_pose) = frame_utils.parse_range_image_and_camera_projection(frame)
```
i prefer to use the [`black`](https://github.com/psf/black) style for multi-line parameter lists:

Here's an example:
https://github.com/mseg-dataset/mseg-semantic/blob/master/mseg_semantic/tool/accuracy_calculator.py#L60

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -107,22 +135,36 @@ def main(save_images: bool, save_poses: bool, save_calibration: bool):
+			(range_images,
+			camera_projections,
+			range_image_top_pose) = frame_utils.parse_range_image_and_camera_projection(frame)
+			points_ri2, cp_points_ri2 = frame_utils.convert_range_image_to_point_cloud(
```
is this using the first return?

hchittanuru3 replied:
I changed it to use first return by default, but now the user can specify using `argparse`.

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -216,19 +258,89 @@ def dump_pose(city_SE3_egovehicle, timestamp, log_id):
 	check_mkdir(str(Path(json_fpath).parent))
 	save_json_dict(json_fpath, pose_dict)
 
+def dump_point_cloud(points, timestamp, log_id, parent_path):
```
thanks for adding this. do you mind adding a 1-line docstring and type hints here?

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -216,19 +258,89 @@ def dump_pose(city_SE3_egovehicle, timestamp, log_id):
+	check_mkdir(str(Path(json_fpath).parent))
+	save_json_dict(json_fpath, argoverse_labels)
+
+def build_label(label, timestamp):
```
maybe we could name this `build_argo_label()` or `convert_waymo_to_argo_label()` to be a bit more descriptive?

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -216,19 +258,89 @@ def dump_pose(city_SE3_egovehicle, timestamp, log_id):
+	label_dict["timestamp"] = timestamp
+	if label.id not in track_id_dict.keys():
+		track_id = uuid.uuid4().hex
+		track_id_dict[label.id] = track_id
```
is `track_id_dict` a global variable? I'd prefer it not be

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -45,6 +52,14 @@
     'ring_side_right', # 5, 'SIDE_RIGHT'
 ]
 
+LABEL_TYPES = [
```
could you add an in-line comment explaining what the indices here denote? which ones are argo, and which one waymo labels?

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/8931614c713c462a1efbbc0f06a366d44a28df25)

johnwlambert reviewed (changes requested):
Looks good, Hemanth. Thanks for making this. Two other comments:

(1) are we modifying the calibration files before saving them, with those lines i showed?
(2) do you mind adding to the installation instructions any snags you ran into?

Commit:
Adding distortion and camera projection

hchittanuru3 commented:
> Looks good, Hemanth. Thanks for making this. Two other comments:
> 
> (1) are we modifying the calibration files before saving them, with those lines i showed?
> (2) do you mind adding to the installation instructions any snags you ran into?

I added the calibration code, but when generating images to test, the boxes aren't showing up, so need to debug that. I added some installation and running instructions as well.

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    files = glob.glob(f"{record_dir}/*.tfrecord")
+    log_ids = []
+    for file in files:
+        file = file.strip(record_dir)
```
`strip()` can do some weird things, like 

```
>>> mystr = '/hello/world/its/me'
>>> mystr.strip('/hello/world/')
'its/m'
>>> mystr
'/hello/world/its/me'
>>> mystr.strip('/hello/world/its/')
'm'
```
Would `file.replace(record_dir, "")` be better?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/9db2006d2741dbd66d59c22e07604a9bd62fa2a6)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+            if args.save_poses:
+                dump_pose(city_SE3_egovehicle, timestamp_ns, log_id, ARGO_WRITE_DIR)
+            # Reading lidar data and saving it in point cloud format
+            # We are only using the first range image by default (Waymo provides two range images)
```
i think these lines are 99 chars long -- can we break them into 80-char length lines?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/9db2006d2741dbd66d59c22e07604a9bd62fa2a6)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+
+
+def undistort_image(img: np.ndarray, calib_data: Any, camera_name: int):
+    """Undistort the image from the Waymo dataset given camera calibration data"""
```
do you mind briefly mentioning what k1, k2, k3 and p1, p2 represent here? (tangential vs. radial)?

johnwlambert replied:
can we add a brief unit test here as well, suppose we start with a 40x40 white image with a black diagonal from bottom-left to top-right, and top-left to bottom-right? THen we distort it ourselves, and make sure we can recover a clean "X" when we call this function?

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    """Saves the SE3 transformation from city frame to egovehicle frame at a particular timestamp
+
+    Args:
+        city_SE3_egovehicle: A (4,4) numpy array representing the SE3 transformation from city to egovehicle frame
```
115 char line, can we break down to 80 chars?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/9db2006d2741dbd66d59c22e07604a9bd62fa2a6)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    save_json_dict(json_fpath, pose_dict)
+
+
+def dump_point_cloud(
```
can we add a unit test on this function? It can use `argoverse.utils.ply_reader import read_ply` to just make sure we get back what we save

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/3b5f39632d27d0d05165fd850e73bd772ceaaaf5)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    argoverse_labels = []
+    for label in labels:
+        argoverse_labels.append(build_argo_label(label, timestamp, track_id_dict))
+    json_fpath = f"{parent_path}/{log_id}/per_sweep_annotations_amodal/tracked_object_labels_{timestamp}.json"
```
111 chars, can we do this instead:

```
json_fpath = f"{parent_path}/{log_id}/per_sweep_annotations_amodal/"
json_fpath += f"tracked_object_labels_{timestamp}.json"
``

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/5aa62285b6622c6372a097bc05a24214f4f59757)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    save_json_dict(json_fpath, argoverse_labels)
+
+
+def build_argo_label(
```
can we add a simple unit test/smokescreen on this function, just to make sure it runs without crashing?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/9db2006d2741dbd66d59c22e07604a9bd62fa2a6)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+        help="whether to save object labels or not",
+    )
+    parser.add_argument(
+        "--range-image",
```
are we using this parameter in the code? i couldn't find it with ctrl-f

johnwlambert replied:
nevermind, it is automatically converted in argparse i guess

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+                ) = frame_utils.convert_range_image_to_point_cloud(
+                    frame, range_images, camera_projections, range_image_top_pose
+                )
+            elif args.range_image == 2:
```
i think we had a hyphen, not underscore, in the argparse, between range-image?

johnwlambert replied:
nevermind, they are automatically converted in argparse i guess

hchittanuru3 replied:
I think `argparse` converts the hyphenated arguments to underscores? You can see an example [here](https://github.com/argoai/argoverse-api/blob/master/demo_usage/cuboids_to_bboxes.py#L258)

johnwlambert replied:
good point, you're right

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    ARGO_WRITE_DIR = args.argo_dir
+    track_id_dict = {}
+    img_count = 0
+    log_ids = get_log_id_from_files(TFRECORD_DIR)[:1]
```
why do we need the `[:1]` index here?

hchittanuru3 replied:
Good catch, I was doing this for testing.

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+                range_image_top_pose,
+            ) = frame_utils.parse_range_image_and_camera_projection(frame)
+            if args.range_image == 1:
+                (
```
i would prefer without the parentheses on the return args, so the following instead:

````python
points_ri, cp_points_ri = frame_utils.convert_range_image_to_point_cloud(
    frame, range_images, camera_projections, range_image_top_pose
)
```

hchittanuru3 replied:
Sorry `black` was doing this. Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/3b5f39632d27d0d05165fd850e73bd772ceaaaf5)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+                        frame.context.camera_calibrations,
+                        tf_cam_image.name,
+                    )
+                    img_save_fpath = f"{ARGO_WRITE_DIR}/{log_id}/{camera_name}/{camera_name}_{timestamp_ns}.jpg"
```
do we need to round any of these nanosecond timestamps to nearest microseconds? (i.e. 
`t_nanos = 1508103378165379072` becomes `t_micros = 1508103378165379000`? I remember there were weird mismatches when I did this, because of different time types/rounding/precision on their end. We should check if labels and lidar and camera times are all provided in nanoseconds

hchittanuru3 replied:
Yeah, that's a good point. Should we just round the timestamp from the camera image to microseconds so we don't have mismatches?

johnwlambert replied:
Hmm maybe it's ok? Let's see if the LiDAR visualization on images works ok with the current format:

Camera pose timestamp and image timestamp are both doubles: https://github.com/waymo-research/waymo-open-dataset/blob/master/waymo_open_dataset/dataset.proto#L244. They must match for the Argoverse tools to work.

The LiDAR timestamp is a int64 https://github.com/waymo-research/waymo-open-dataset/blob/master/waymo_open_dataset/dataset.proto#L298 
but it's fine if that doesn't match the LiDAR, since it's an entirely different sensor.

hchittanuru3 replied:
Yeah so the camera pose and image timestamp are always the same. I saw that `cuboids_to_bboxes` from argoverse finds the closest LIDAR timestamp.

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    """Get the log IDs of the Waymo records from the directory where they are stored
+
+    Args:
+        record_dir: The path to the directory where the Waymo data is stored
```
can we add an example of what one of their `record_dir` strings looks like (just an example?)

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/5aa62285b6622c6372a097bc05a24214f4f59757)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+    )
+    parser.add_argument(
+        "--save-calibration",
+        default=True,
```
in `argparse` the bool type won't allow a False declaration ever (will always be True). Try running this: `python test_argparse_bool.py --save-images False --range-image 2`
```python
import argparse
import pdb

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--save-images", default=True, type=bool, help="whether to save images or not"
    )

    parser.add_argument(
        "--range-image",
        default=1,
        type=int,
        choices=[1, 2],
        help="which range image to use from Waymo",
    )
    args = parser.parse_args()
    pdb.set_trace()
    main(args)
```
We need to use `store_false` or `store_true` like https://github.com/mseg-dataset/mseg-api/blob/master/mseg/label_preparation/remap_dataset.py#L190

hchittanuru3 replied:
Good catch! I used `str2bool` (addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/5aa62285b6622c6372a097bc05a24214f4f59757))

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+        standardcam_SE3_egovehicle = standardcam_SE3_waymocam.right_multiply_with_se3(
+            waymocam_SE3_egovehicle
+        )
+        egovehicle_q_camera = rotmat2quat(standardcam_SE3_egovehicle.rotation)
```
this line is wrong, we need an inverse first:

```python
egovehicle_SE3_standardcam = standardcam_SE3_egovehicle.inverse()
egovehicle_q_camera = rotmat2quat(egovehicle_SE3_standardcam.rotation)
```

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/5aa62285b6622c6372a097bc05a24214f4f59757)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
+            waymocam_SE3_egovehicle
+        )
+        egovehicle_q_camera = rotmat2quat(standardcam_SE3_egovehicle.rotation)
+        x, y, z = standardcam_SE3_egovehicle.translation
```
need to take x,y,z from the inverse, since on disk we are saving `"vehicle_SE3_camera_"`, not `camera_SE3_vehicle` (see line below)

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/5aa62285b6622c6372a097bc05a24214f4f59757)

johnwlambert commented on waymo_raw_data_to_argoverse.py:
```diff
@@ -30,205 +40,398 @@
-	save_calibration = True
-
-	main(save_images, save_poses, save_calibration)
+def save_json_dict(
```
we can remove this function and just import 
```python
from argoverse.utils.json_utils import save_json_dict
```

hchittanuru3 replied:
Addressed in [this commit](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/5aa62285b6622c6372a097bc05a24214f4f59757)

Commit:
Refactoring and addressing matrix inverses

Commit:
Added tests

Commit:
Updating reqs

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,444 @@
+    """
+    argoverse_labels = []
+    for label in labels:
+        if label.type != 3:
```
thanks for updating this Hemanth. Can we make it `label.type != LABEL_TYPES.index("SIGN")` to make this explicit, and then a brief inline comment explaining we don't want SIGN, since not a category in Argoverse?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/3b5f39632d27d0d05165fd850e73bd772ceaaaf5)

Commit:
Point cloud test done

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,441 @@
+            return cv2.undistort(img, camera_matrix, dist_coeffs)
+
+
+def form_calibration_json(calib_data):
```
can we add the types for input and return arguments here?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/4ce61dbfa76327be7e9627546897df849c7b8cff)

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,441 @@
+    Args:
+        record_dir: The path to the directory where the Waymo data
+                    is stored
+                    Example: "/path-to-waymo-data"
```
i was thinking something a bit more like
`Example: segment-8806931859563747931_1160_000_1180_000.tfrecord`
that's the standard file name format, right?

hchittanuru3 replied:
Yeah that's the file format, but `record_dir` is the path to the directory that has all these files.

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,441 @@
+        log_id: Log ID that the reading belongs to
+        parent_path: The directory that the converted data is written to
+    """
+    points = points.astype(float)
```
maybe we can add a small inline comment just mentioning that it must be float since integer arrays cause problems

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/4ce61dbfa76327be7e9627546897df849c7b8cff)

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,441 @@
+                        print(f"\tSaved {img_count}'th image for log = {log_id}")
+
+
+def undistort_image(img: np.ndarray, calib_data: Any, camera_name: int):
```
i think missing return type here

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/4ce61dbfa76327be7e9627546897df849c7b8cff)

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,441 @@
+
+def undistort_image(img: np.ndarray, calib_data: Any, camera_name: int):
+    """Undistort the image from the Waymo dataset given camera calibration data"""
+    for camera_calib in calib_data:
```
I wonder if we can get rid of this for loop -- if camera_name is an integer, is it an index into this list? if not, maybe we could just store calibration in a dictionary instead

hchittanuru3 replied:
So it's like indexing into a list, and I wasn't able to index it by camera name. I also tried serializing it but we might have to serialize the whole frame, and not use the waymo open dataset's function `open_dataset.Frame()`.

https://github.com/googleapis/google-cloud-python/issues/3485

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,441 @@
+# Mapping from Argo Label types to Waymo Label types
+# The indices correspond to Waymo's label types
+LABEL_TYPES = [
+    "OTHER_MOVER",  # 0, TYPE_UNKNOWN
```
I'm still a bit confused about this part.

Maybe we can add one more inline comment saying Waymo labels are on the right, and
corresponding Argoverse labels are on the left, with a link to https://github.com/argoai/argoverse-api/blob/master/argoverse/data_loading/object_classes.py#L6

Or are there just 5 label types in Waymo Open Dataset?

I'm not sure that Waymo's `UNKNOWN` would correspond to Argoverse' `OTHER_MOVER`?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/4ce61dbfa76327be7e9627546897df849c7b8cff)

Yeah according to their GitHub, there are 5 label types.

johnwlambert commented on waymo2argo/waymo_raw_data_to_argoverse.py:
```diff
@@ -0,0 +1,441 @@
+        os.makedirs(dirpath, exist_ok=True)
+
+
+def get_log_id_from_files(record_dir: str) -> List[str]:
```
maybe we can name this `get_log_ids_from_files()` since we are returning more than one `log_id`?

hchittanuru3 replied:
Addressed [here](https://github.com/johnwlambert/waymo_to_argoverse/pull/2/commits/4ce61dbfa76327be7e9627546897df849c7b8cff)

Commit:
Cosmetic changes

johnwlambert committed:
reorder imports and fix type hint

johnwlambert committed:
clean up the docstring

johnwlambert committed:
make docstring style a bit more consistent

johnwlambert committed:
clean up type hints

johnwlambert committed:
allow easier visual diff in ui

johnwlambert committed:
delete init, we can avoid it using module installation

johnwlambert committed:
clean up a few docstrings, put argparse code next to each other

johnwlambert commented:
Thanks for the @hchittanuru3, it looks good. I'm going ahead and merging it now.
